ethereum/core/src/main/java/org/hyperledger/besu/ethereum/encoding/ProtocolRLPSpec.java

```diff
@@ -0,0 +1,230 @@
+/*
+ *
+ *  * Copyright ConsenSys AG.
+ *  *
+ *  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ *  * the License. You may obtain a copy of the License at
+ *  *
+ *  * http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ *  * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *  * specific language governing permissions and limitations under the License.
+ *  *
+ *  * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package org.hyperledger.besu.ethereum.encoding;
+
+import org.hyperledger.besu.config.experimental.ExperimentalEIPs;
+import org.hyperledger.besu.ethereum.core.Address;
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.core.BlockBody;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.BlockHeaderFunctions;
+import org.hyperledger.besu.ethereum.core.Difficulty;
+import org.hyperledger.besu.ethereum.core.Hash;
+import org.hyperledger.besu.ethereum.core.LogsBloomFilter;
+import org.hyperledger.besu.ethereum.core.Transaction;
+import org.hyperledger.besu.ethereum.core.Wei;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.ethereum.rlp.RLP;
+import org.hyperledger.besu.ethereum.rlp.RLPInput;
+import org.hyperledger.besu.ethereum.rlp.RLPOutput;
+import org.hyperledger.besu.plugin.data.Quantity;
+import org.hyperledger.besu.plugin.data.TransactionType;
+
+import java.util.List;
+import java.util.Optional;
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.tuweni.bytes.Bytes;
+
+public interface ProtocolRLPSpec {
+
+  @FunctionalInterface
+  interface Encoder<T> {
+    void encode(T object, RLPOutput output);
+  }
+
+  @FunctionalInterface
+  interface Decoder<T> {
+    T decode(RLPInput input);
+  }
+
+  ImmutableMap<TransactionType, Encoder<Transaction>> TYPED_TRANSACTION_ENCODERS =
+      ImmutableMap.of(TransactionType.EIP1559, ProtocolRLPSpec::encodeEIP1559);
+
+  // This is all going to change in the next PR because we had to settle some spec ambiguities on a
+  // breakout room call
+  static void encode(final Transaction transaction, final RLPOutput rlpOutput) {
```

Comment:
I don't like how the `writeTo` and `readFrom` methods were removed from the associated data objects.  It forces these encoding objects to access what should be encapsulated data.  

It also forces us to do weird lambda dances, where `foo::writeTo` becomes `rlpOut -> Static.method(foo, rlpOut)`.  This transform makes me thing we did something wrong.